fw/mmap_buffer.c

```diff
@@ -0,0 +1,329 @@
+	else
+		return -EINVAL;
+
+#define NTH_ONLINE_CPU(n) ({		\
+	int cpu, res = -1, i = 0;	\
+	for_each_online_cpu(cpu) {	\
+		if (i == n) {		\
+			res = cpu;	\
+			break;		\
+		}			\
+		++i;			\
+	}				\
+	res;				\
+})
+
+	cpu_num = vma->vm_pgoff / (area_size / PAGE_SIZE);
+	cpu_id = NTH_ONLINE_CPU(cpu_num);
+	if (cpu_id < 0)
+		return -EINVAL;
+
+	buf = *per_cpu_ptr(holder->buf, cpu_id);
+
+	pfn = page_to_pfn(holder->mem[cpu_id].buf_page);
+	if (remap_pfn_range(vma, vma->vm_start, pfn,
+			    TFW_MMAP_BUFFER_DATA_OFFSET, vma->vm_page_prot))
+		return -EAGAIN;
+
+	if (area_size == FULL_SIZE) { /* entire buffer is mapping */
+		pfn = page_to_pfn(holder->mem[cpu_id].data_page);
+		user_addr = vma->vm_start + TFW_MMAP_BUFFER_DATA_OFFSET;
+		for (i = 0; i < 2; ++i) {
+			if (remap_pfn_range(vma, user_addr, pfn, buf->size,
+					    vma->vm_page_prot)) {
+				return -EAGAIN;
+			}
+			user_addr += buf->size;
+		}
+
+		vma->vm_ops = &dev_vm_ops;
+		atomic_set(&buf->is_ready, 1);
+	}
+
+	return 0;
+
+#undef NTH_ONLINE_CPU
+#undef FULL_SIZE
+}
+
+static void
+dev_file_vm_close(struct vm_area_struct *vma)
+{
+	TfwMmapBufferHolder *holder = vma->vm_file->private_data;
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	atomic_set(&buf->is_ready, 0);
+}
+
+TfwMmapBufferHolder *
+tfw_mmap_buffer_create(const char *filename, unsigned int size)
+{
+	TfwMmapBufferHolder *holder;
+	struct page **page_ptr = NULL;
+	unsigned int order, i, page_cnt;
+	int cpu;
+
+	if (size < TFW_MMAP_BUFFER_MIN_SIZE
+	    || size > TFW_MMAP_BUFFER_MAX_SIZE
+	    || !is_power_of_2(size))
+		return NULL;
+
+	if (filename && strlen(filename) >= TFW_MMAP_BUFFER_MAX_NAME_LEN - 1)
+		return NULL;
+
+	if (holders_cnt + 1 > MAX_HOLDERS)
+		return NULL;
+
+	holder = kzalloc(sizeof(TfwMmapBufferHolder) +
+			 sizeof(TfwMMapBufferMem) * num_online_cpus(),
+			 GFP_KERNEL);
+	if (!holder)
+		return NULL;
+
+	order = get_order(size);
+
+	holder->dev_major = -1;
+	holder->size = size;
+	holder->buf = (TfwMmapBuffer **)alloc_percpu_gfp(sizeof(TfwMmapBuffer *),
+							 GFP_KERNEL);
+	if (!holder->buf)
+		goto err;
+
+	atomic_set(&holder->dev_is_opened, 0);
+	atomic_set(&holder->is_freeing, 0);
+
+	page_cnt = size / PAGE_SIZE;
+	/*
+	 * Allocate pages for double mapping and a page for buffer control structure.
+	 */
+	page_ptr = kmalloc((page_cnt * 2 + 1) * sizeof(struct page *), GFP_KERNEL);
+	if (!page_ptr)
+		goto err;
+
+	for_each_online_cpu(cpu) {
+		TfwMmapBuffer *buf, **bufp;
+
+		holder->mem[cpu].buf_page = alloc_pages_node(cpu_to_node(cpu),
+							     GFP_KERNEL, 0);
+		if (holder->mem[cpu].buf_page == NULL)
+			goto err;
+
+		holder->mem[cpu].data_page = alloc_pages_node(cpu_to_node(cpu),
+							      GFP_KERNEL, order);
+		if (holder->mem[cpu].data_page == NULL)
+			goto err;
+
+		page_ptr[0] = holder->mem[cpu].buf_page;
+		for (i = 0; i < page_cnt; ++i) {
+			page_ptr[i + 1] = &holder->mem[cpu].data_page[i];
+			page_ptr[i + page_cnt + 1] = &holder->mem[cpu].data_page[i];
```

Comment:
why `page_ptr[i + page_cnt + 1] = &holder->mem[cpu].data_page[i];` and `page_ptr[i + 1] = &holder->mem[cpu].data_page[i];` (same value)?

Reply:
When we have ringbuffer data can be sliced to 2 parts:
```
-----------            -----------
| >> data |            | data >> |
----------------------------------
|               buf              |
----------------------------------
```
We map the buffer twice. In such a way we get rid of bound issues and read data in one piece:
```
                    ------------
                    |   data   |
------------------------------------------
| header |      buf      |      buf      |
------------------------------------------

         ^               ^
         |               |
	/                 \
       /                   \
     +1                     (+1 + page_cnt)
```